### ISSUE TITLE
Could not get response body by using response.text(), response.buffer() or response.json() methods

### ISSUE THREAD

**Issue by snakesgun** [puppeteer/puppeteer#2176]
### Steps to reproduce

My environment is as follow:

* Puppeteer version: 1.1.1
* Platform / OS version: OS X 10.13.3
* URLs (if applicable): https://fanyi.baidu.com/#en/zh/hello
* Node.js version: v9.8.0

**To re-produce the problem**

``` javascript
const puppeteer = require('puppeteer');
(async () => {
	const browser = await puppeteer.launch();
  const page = await browser.newPage();
  console.log("test");
	page.on('response', response => {
		const req = response.request();
		if (req.url() === 'https://fanyi.baidu.com/v2transapi') {
			response.buffer().then(
				b => {
					console.log(`${response.status()} ${response.url()} ${b.length} bytes`);
				},
				e => {
					console.error(`${response.status()} ${response.url()} failed: ${e}`);
				}
			);
		}
	});

	await page.goto('https://fanyi.baidu.com/#en/zh/hello');
	await page.close();
	await browser.close();
})();
```
**What is the expected result?**
``` bash
200 https://fanyi.baidu.com/v2transapi 179360 bytes
```
**What happens instead?**
``` bash
200 https://fanyi.baidu.com/v2transapi failed: Error: Protocol error (Network.getResponseBody): No resource with given identifier found undefined
```

**Comment by dinigo** [comment]
I suspect this might be related to https://github.com/GoogleChrome/puppeteer/issues/2258 and this https://github.com/GoogleChrome/puppeteer/issues/2284 where @aslushnikov explains this is due a race condition.

Just doing some archeology because I am experiencing this same problem.

**Comment by dinigo** [comment]
@snakesgun , I just run this without any error.
```javascript
const puppeteer = require('puppeteer');
(async () => {
  const browser = await puppeteer.launch();
  const page = await browser.newPage();
  console.log('Starting navigation');
  page.on('response', async response => {
    const url = response.url();
    try {
      const req = response.request();
      const orig = req.url();
      const text = await response.text();
      const status = response.status();
      console.log({orig, status, text: text.length});
    } catch (err) {
      console.error(`Failed getting data from: ${url}`);
      console.error(err);
    }
  });
  await page.goto('https://fanyi.baidu.com/#en/zh/hello');
  await page.close();
  await browser.close();
})();
```

I'd rather had it run with an error because I would know that something fails here. I suppose mi code breaks somewhere else.

**Comment by dinigo** [comment]
Hi, I just debugged my error. I was trying to get the response _body_ from responses which doesn't have any like
- Pixels calls (for analytics)
- 3XX: redirects
- 204: all OK but no content response

```javascript
let status;
let text;
if(response.status) {
  status = response.status();
}
if(
  status // we actually have an status for the response
  && !(status > 299 && status < 400) // not a redirect
  && !(status === 204) // not a no-content response
  && !(resourceType === 'image') // not an image
) {
  text =  await response.text();
}
```

I'm leaving this here for reference

**Comment by nadzimo** [comment]
I'm also having the same problem as OP. The response status is 200 and I can't get the buffer from the response with `const buff = await response.buffer()`, also getting the error `UnhandledPromiseRejectionWarning: Error: Protocol error (Network.getResponseBody): No resource with given identifier found`
Has there been progress on this issue?

**Comment by vsemozhetbyt** [comment]
@nadzimo Can you provide a small script to reproduce?

**Comment by nadzimo** [comment]
> @nadzimo Can you provide a small script to reproduce?

I actually just noticed that it's working fine on non-file download requests (e.g. fetching css, img), but it's failing when the request is a file download.  
Here is the response header when the browser is requesting that file download (I sanitized it for security reasons):  

```
Cache-Control: no-cache, no-store, must-revalidate
Connection: Keep-Alive
Content-Disposition: attachment; filename=582148_1550843116740.CSV
Content-Encoding: gzip
Content-Length: 3458
Content-Type: text/csv
Date: Fri, 22 Feb 2019 13:11:50 GMT
Expires: Thu, 01 Jan 1970 00:00:00 GMT
Keep-Alive: timeout=15
Pragma: no-cache
Set-Cookie: TS01b4a281=016120891722f0710f93c20268ae3cca4d6d353f25ad9d9ecb46899c3927d9d9f32916405b13b266d0135623c04d48a95b94abe1de0dee30a224712aca7c21dece5b7b2f5041841cf9eb1bd285a17fb877b36a8; Path=/; Domain=.xxx.xxx.com
Strict-Transport-Security: max-age=15552000; includeSubDomains
X-Content-Type-Options: 'nosniff'
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block
```

**Comment by aslushnikov** [comment]
@snakesgun turns out the page in question reloads itself with javascript before we have time to load the body for the response.

Consider the following script:

```js
const puppeteer = require('puppeteer');
const API_URL = 'https://fanyi.baidu.com/v2transapi';
(async () => {
  const browser = await puppeteer.launch();
  const page = await browser.newPage();
  page.on('response', r => {
    if (r.request().isNavigationRequest() && r.frame() === page.mainFrame())
      console.log(`== NAVIGATION COMMITTED TO ${r.url()} ==`);
  });

  page.on('requestfinished', async req => {
    if (req.url() !== API_URL)
      return;
    const response = req.response();
    console.log(`- fetching: ${API_URL}`);
    try {
      const buffer= await response.buffer();
      console.log(`- done: ${buffer.length} bytes`);
    } catch (e) {
      console.error(`- failed: ${e}`);
    }
  });

  await page.goto('https://fanyi.baidu.com/#en/zh/hello');
})();
```

The output I'm getting is:
```bash
lushnikov:~/prog/puppeteer(master)$ node a.mjs
== NAVIGATION COMMITTED TO https://fanyi.baidu.com/ ==
- fetching: https://fanyi.baidu.com/v2transapi
== NAVIGATION COMMITTED TO https://fanyi.baidu.com/ ==
- failed: Error: Protocol error (Network.getResponseBody): No resource with given identifier found
- fetching: https://fanyi.baidu.com/v2transapi
- done: 64411 bytes
```

Notice:
- there are actually two requests to the `v2..` endpoint
- the script I attached doesn't include `page.close` and `browser.close`. These were triggered too early since by default, `page.goto` waits til `load` event, and it happens before the javascript redirect is processed.

To sum up: this is not a bug; such things might happen and should be handled gracefully.
